### PR TITLE
Correctly handle links starting with whitespace

### LINKED ISSUE #107995: Rustdoc Internal Compiler Error for Diesel
### Code

```Rust
git clone https://github.com/diesel-rs/diesel
cd diesel 
git checkout 0c8a87713 # current master branch
cargo doc --manifest-path diesel/Cargo.toml --features "postgres sqlite mysql extras i-implement-a-third-party-backend-and-opt-into-breaking-changes" --workspace

```

### Meta

`rustc --version --verbose`:
```
rustc 1.69.0-nightly (5b8f28453 2023-02-12)
```

### Error output

```
thread 'rustc' panicked at 'no resolution for "LoadRowIter" MacroNS DefId(0:181 ~ diesel[46ea]::connection)', src/librustdoc/passes/collect_intra_doc_links.rs:393:32
stack backtrace:
   0:     0x7fdfa9b8803a - std::backtrace_rs::backtrace::libunwind::trace::h68819555df8b2d9c
                               at /rustc/5b8f284536d00ba649ca968584bedab4820d8527/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x7fdfa9b8803a - std::backtrace_rs::backtrace::trace_unsynchronized::hd69f04a0acc016b0
                               at /rustc/5b8f284536d00ba649ca968584bedab4820d8527/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7fdfa9b8803a - std::sys_common::backtrace::_print_fmt::hd05f5f45b3abc36d
                               at /rustc/5b8f284536d00ba649ca968584bedab4820d8527/library/std/src/sys_common/backtrace.rs:65:5
   3:     0x7fdfa9b8803a - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h97505bc956055c86
                               at /rustc/5b8f284536d00ba649ca968584bedab4820d8527/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x7fdfa9be821e - core::fmt::write::hee40881876785ec8
                               at /rustc/5b8f284536d00ba649ca968584bedab4820d8527/library/core/src/fmt/mod.rs:1232:17
   5:     0x7fdfa9b78755 - std::io::Write::write_fmt::h9ac4fb15bcdd983e
                               at /rustc/5b8f284536d00ba649ca968584bedab4820d8527/library/std/src/io/mod.rs:1684:15
   6:     0x7fdfa9b87e05 - std::sys_common::backtrace::_print::h7b79888859e1041b
                               at /rustc/5b8f284536d00ba649ca968584bedab4820d8527/library/std/src/sys_common/backtrace.rs:47:5
   7:     0x7fdfa9b87e05 - std::sys_common::backtrace::print::h5f377a03e09d721c
                               at /rustc/5b8f284536d00ba649ca968584bedab4820d8527/library/std/src/sys_common/backtrace.rs:34:9
   8:     0x7fdfa9b8ab9f - std::panicking::default_hook::{{closure}}::h081dc35da89fae25
                               at /rustc/5b8f284536d00ba649ca968584bedab4820d8527/library/std/src/panicking.rs:267:22
   9:     0x7fdfa9b8a8db - std::panicking::default_hook::h5ac9a56440f0f320
                               at /rustc/5b8f284536d00ba649ca968584bedab4820d8527/library/std/src/panicking.rs:286:9
  10:     0x7fdfacee4144 - rustc_driver_impl[45704de6fc1a7e42]::DEFAULT_HOOK::{closure#0}::{closure#0}
  11:     0x7fdfa9b8b3da - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::h597f484af91b69d8
                               at /rustc/5b8f284536d00ba649ca968584bedab4820d8527/library/alloc/src/boxed.rs:2002:9
  12:     0x7fdfa9b8b3da - std::panicking::rust_panic_with_hook::h32591b11d4f9c8c2
                               at /rustc/5b8f284536d00ba649ca968584bedab4820d8527/library/std/src/panicking.rs:692:13
  13:     0x7fdfa9b8b159 - std::panicking::begin_panic_handler::{{closure}}::hb2b75e52ec6963d4
                               at /rustc/5b8f284536d00ba649ca968584bedab4820d8527/library/std/src/panicking.rs:579:13
  14:     0x7fdfa9b884dc - std::sys_common::backtrace::__rust_end_short_backtrace::ha859a147e693367c
                               at /rustc/5b8f284536d00ba649ca968584bedab4820d8527/library/std/src/sys_common/backtrace.rs:137:18
  15:     0x7fdfa9b8ae62 - rust_begin_unwind
                               at /rustc/5b8f284536d00ba649ca968584bedab4820d8527/library/std/src/panicking.rs:575:5
  16:     0x7fdfa9be4bc3 - core::panicking::panic_fmt::hb9e85c5c10f5beb5
                               at /rustc/5b8f284536d00ba649ca968584bedab4820d8527/library/core/src/panicking.rs:64:14
  17:     0x5592c323b52f - <rustdoc[e05275ab24b811c1]::passes::collect_intra_doc_links::LinkCollector>::resolve_path
  18:     0x5592c323b662 - <rustdoc[e05275ab24b811c1]::passes::collect_intra_doc_links::LinkCollector>::resolve
  19:     0x5592c32401a9 - <rustdoc[e05275ab24b811c1]::passes::collect_intra_doc_links::LinkCollector as rustdoc[e05275ab24b811c1]::visit::DocVisitor>::visit_item
  20:     0x5592c324d46a - <rustdoc[e05275ab24b811c1]::passes::collect_intra_doc_links::LinkCollector as rustdoc[e05275ab24b811c1]::visit::DocVisitor>::visit_inner_recur
  21:     0x5592c3242f1f - <rustdoc[e05275ab24b811c1]::passes::collect_intra_doc_links::LinkCollector as rustdoc[e05275ab24b811c1]::visit::DocVisitor>::visit_item
  22:     0x5592c324d42a - <rustdoc[e05275ab24b811c1]::passes::collect_intra_doc_links::LinkCollector as rustdoc[e05275ab24b811c1]::visit::DocVisitor>::visit_inner_recur
  23:     0x5592c3242f05 - <rustdoc[e05275ab24b811c1]::passes::collect_intra_doc_links::LinkCollector as rustdoc[e05275ab24b811c1]::visit::DocVisitor>::visit_item
  24:     0x5592c324d42a - <rustdoc[e05275ab24b811c1]::passes::collect_intra_doc_links::LinkCollector as rustdoc[e05275ab24b811c1]::visit::DocVisitor>::visit_inner_recur
  25:     0x5592c3242f05 - <rustdoc[e05275ab24b811c1]::passes::collect_intra_doc_links::LinkCollector as rustdoc[e05275ab24b811c1]::visit::DocVisitor>::visit_item
  26:     0x5592c3239fbf - rustdoc[e05275ab24b811c1]::passes::collect_intra_doc_links::collect_intra_doc_links
  27:     0x5592c31fa03e - <rustc_session[15418ad64be04b7e]::session::Session>::time::<rustdoc[e05275ab24b811c1]::clean::types::Crate, rustdoc[e05275ab24b811c1]::core::run_global_ctxt::{closure#7}>
  28:     0x5592c3277c5b - rustdoc[e05275ab24b811c1]::core::run_global_ctxt
  29:     0x5592c31fa32b - <rustc_session[15418ad64be04b7e]::session::Session>::time::<(rustdoc[e05275ab24b811c1]::clean::types::Crate, rustdoc[e05275ab24b811c1]::config::RenderOptions, rustdoc[e05275ab24b811c1]::formats::cache::Cache), rustdoc[e05275ab24b811c1]::main_args::{closure#1}::{closure#0}::{closure#1}::{closure#0}>
  30:     0x5592c303c71d - <rustc_interface[1dae503a51862634]::queries::QueryResult<&rustc_middle[950e6c013232a13e]::ty::context::GlobalCtxt>>::enter::<core[8fc858d2dd58175c]::result::Result<(), rustc_errors[9a7f2c558b529898]::ErrorGuaranteed>, rustdoc[e05275ab24b811c1]::main_args::{closure#1}::{closure#0}::{closure#1}>
  31:     0x5592c2fb2c8a - <rustc_interface[1dae503a51862634]::interface::Compiler>::enter::<rustdoc[e05275ab24b811c1]::main_args::{closure#1}::{closure#0}, core[8fc858d2dd58175c]::result::Result<(), rustc_errors[9a7f2c558b529898]::ErrorGuaranteed>>
  32:     0x5592c31f972b - rustc_span[9e92ed5877933bb8]::with_source_map::<core[8fc858d2dd58175c]::result::Result<(), rustc_errors[9a7f2c558b529898]::ErrorGuaranteed>, rustc_interface[1dae503a51862634]::interface::run_compiler<core[8fc858d2dd58175c]::result::Result<(), rustc_errors[9a7f2c558b529898]::ErrorGuaranteed>, rustdoc[e05275ab24b811c1]::main_args::{closure#1}>::{closure#0}::{closure#0}>
  33:     0x5592c3147462 - <scoped_tls[d2579b36b1a9f18d]::ScopedKey<rustc_span[9e92ed5877933bb8]::SessionGlobals>>::set::<rustc_interface[1dae503a51862634]::interface::run_compiler<core[8fc858d2dd58175c]::result::Result<(), rustc_errors[9a7f2c558b529898]::ErrorGuaranteed>, rustdoc[e05275ab24b811c1]::main_args::{closure#1}>::{closure#0}, core[8fc858d2dd58175c]::result::Result<(), rustc_errors[9a7f2c558b529898]::ErrorGuaranteed>>
  34:     0x5592c320dba0 - std[6e58747c831eadda]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[1dae503a51862634]::util::run_in_thread_pool_with_globals<rustc_interface[1dae503a51862634]::interface::run_compiler<core[8fc858d2dd58175c]::result::Result<(), rustc_errors[9a7f2c558b529898]::ErrorGuaranteed>, rustdoc[e05275ab24b811c1]::main_args::{closure#1}>::{closure#0}, core[8fc858d2dd58175c]::result::Result<(), rustc_errors[9a7f2c558b529898]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[8fc858d2dd58175c]::result::Result<(), rustc_errors[9a7f2c558b529898]::ErrorGuaranteed>>
  35:     0x5592c3015b77 - <<std[6e58747c831eadda]::thread::Builder>::spawn_unchecked_<rustc_interface[1dae503a51862634]::util::run_in_thread_pool_with_globals<rustc_interface[1dae503a51862634]::interface::run_compiler<core[8fc858d2dd58175c]::result::Result<(), rustc_errors[9a7f2c558b529898]::ErrorGuaranteed>, rustdoc[e05275ab24b811c1]::main_args::{closure#1}>::{closure#0}, core[8fc858d2dd58175c]::result::Result<(), rustc_errors[9a7f2c558b529898]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[8fc858d2dd58175c]::result::Result<(), rustc_errors[9a7f2c558b529898]::ErrorGuaranteed>>::{closure#1} as core[8fc858d2dd58175c]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  36:     0x7fdfa9b95373 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::hfe75deb0c10a09d3
                               at /rustc/5b8f284536d00ba649ca968584bedab4820d8527/library/alloc/src/boxed.rs:1988:9
  37:     0x7fdfa9b95373 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::hcd4bdd69a317ae2f
                               at /rustc/5b8f284536d00ba649ca968584bedab4820d8527/library/alloc/src/boxed.rs:1988:9
  38:     0x7fdfa9b95373 - std::sys::unix::thread::Thread::new::thread_start::h1aad22be20900f56
                               at /rustc/5b8f284536d00ba649ca968584bedab4820d8527/library/std/src/sys/unix/thread.rs:108:17
  39:     0x7fdfa98b6b43 - <unknown>
  40:     0x7fdfa9948a00 - <unknown>
  41:                0x0 - <unknown>

error: the compiler unexpectedly panicked. this is a bug.
```

<details><summary> cargo bisect-rustc points to:https://github.com/rust-lang/rust/commit/5b450244876154bc1bd134694398e80c12e00b5c

</summary>

searched nightlies: from nightly-2023-01-30 to nightly-2023-02-13
regressed nightly: nightly-2023-02-12
searched commit range: https://github.com/rust-lang/rust/compare/2773383a314a4b8f481ce2bed12c32de794ffbe9...585f3eef26f04440bca726c29193af7b4fa90e54
regressed commit: https://github.com/rust-lang/rust/commit/5b450244876154bc1bd134694398e80c12e00b5c
</details>

@rustbot label +regression-from-stable-to-nightly +T-rustdoc

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes https://github.com/rust-lang/rust/issues/107995.

I just got this issue, wrote a fix and then saw the issue. So here's the PR. ^^'

r? @petrochenkov

### CODE REVIEW

**Comment by petrochenkov** [comment]
I have actually just checked out a new branch for fixing this before going through the mail and seeing this PR.

**Comment by petrochenkov** [comment]
Could you add all the test cases from https://github.com/rust-lang/rust/issues/107995?
```rust
/// [ Path`]
/// [ `Path`]
```
and
```rust
#[doc = "line ["]
#[doc = "Path"]
#[doc = "] line"]
```

File: src/librustdoc/passes/collect_intra_doc_links.rs [hunk 1]
```diff
@@ -897,7 +898,7 @@ fn preprocess_link(
     // Parse and strip the disambiguator from the link, if present.
     let (disambiguator, path_str, link_text) = match Disambiguator::from_str(link) {
         Ok(Some((d, path, link_text))) => (Some(d), path.trim(), link_text.trim()),
-        Ok(None) => (None, link.trim(), link.trim()),
+        Ok(None) => (None, link, link),
```
Comment (petrochenkov) [inline src/librustdoc/passes/collect_intra_doc_links.rs:901]: This is a behavior change, and also an inconsistency with the `compiler/rustc_resolve/src/rustdoc.rs` change.

I'd prefer to decide on what grammar we actually want here (and maybe document it) before making any changes.

Comment (GuillaumeGomez) [reply src/librustdoc/passes/collect_intra_doc_links.rs:901]: This is not. The `trim()` method is called just above.

**Comment by GuillaumeGomez** [comment]
I added the suggested test cases.

File: src/librustdoc/passes/collect_intra_doc_links.rs (hunk 1 above)
Comment (petrochenkov) [reply src/librustdoc/passes/collect_intra_doc_links.rs:901]: It makes a difference whether trim is called before `Disambiguator::from_str` or after.

Comment (GuillaumeGomez) [reply src/librustdoc/passes/collect_intra_doc_links.rs:901]: Well, it's needed for the fix. Where you want the documentation about this to put?

Comment (petrochenkov) [reply src/librustdoc/passes/collect_intra_doc_links.rs:901]: It's not needed for the fix, all the test cases worked before.

I don't want to argue about this too much, it's a reasonable change to support `[ type@Path]`, could you just bring `compiler/rustc_resolve/src/rustdoc.rs` into consistency and do the trim in the same place, between `#` and `@`.

Comment (GuillaumeGomez) [reply src/librustdoc/passes/collect_intra_doc_links.rs:901]: Done!

**Pull request merged**